### PR TITLE
[Enhancement] Make download repos for programming exercises more robust

### LINKED ISSUE #2542: Make download repos for programming exercises more robust
When tutors or instructors download (multiple) repositories of programming exercises, this does not always work correctly.
In case of an exception, it can happen that the Artemis Server does not remove all artifacts making a second download impossible, because the folder and/or the git repository already exists, potentially in a state that cannot be recovered easily.

We should improve this:
1) When a repository is cloned into a subfolder of `repos-download`, we make sure the corresponding folder is empty by removing all entries before the `git clone` operation
2) In case of an exception during the clone operation and any other operation between clone and zip/download, we make sure that the corresponding sub folder is removed securely so that subsequent operations do not produce additional issues

Optional further improvements:
* We tell the user which repositories have been successfully downloaded and which not (e.g. as part of the server response)
* In case of an error, we try to use alternative approaches, e.g. download the zipped repo directly from the Version Control Service (e.g. Bitbucket) in case such an API exists

### PR DESCRIPTION
**Pull request by derLalla**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls

### Motivation and Context

Fixes #2542 

### Description

While cloning into a directory there might be a case, that the selected directory is already existent and not empty. So we delete this directory before cloning into it. 
Also in the case of an error, we delete this directory again. 

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Click the title of the programming exercise
4. Download the students' repositories and verify that everything still works as expected

### Test Coverage

Stays the same, as only two lines of code were changed.

### Screenshots

Click the title "PrTesting2":
image

Click on "Download Repos":
image

### CODE REVIEW

**Review by simonzilker — DISMISSED** [review]
Everything works perfectly on ts1!

**Review by JohannesWeiss — APPROVED** [review]
Tested with @simonzilker, works as expected.

**Review by airelawaleria — DISMISSED** [review]
Tested in the testing session. Works fine

**Review by marvinchimeka — APPROVED** [review]
tested in a testing session. Approve

**Review by DenizShabani — APPROVED** [review]
Tested on TS1 and everything looks good!

**Review by LudwigWeide — APPROVED** [review]
Tested on TS3. Works as intended.

**Review by tiffca — DISMISSED** [review]
Tested successfully on ts1, tried different options for downloading the repo

**Review by valeriebucher — APPROVED** [review]
Tested on ts1, everything worked as described.

**Review by JannisBittlmayer — DISMISSED** [review]
Tested on TS3, works as intended

**Review by FrankeLukas — DISMISSED** [review]
Code changes lgtm.

**Review by FrankeLukas — APPROVED** [review]
Re-approve (only comment was added).

**Review by simonzilker — APPROVED** [review]
re-approve

**Review by JohannesWeiss — APPROVED** [review]
Re-Approve.

**Review by marvinchimeka — APPROVED** [review]
re-approve

**Review by airelawaleria — DISMISSED** [review]
Re-approve

**Review by LudwigWeide — CHANGES_REQUESTED** [review]
Re-approve

**Review by LudwigWeide — DISMISSED** [review]
Re-approve

**Review by FlorianStang — DISMISSED** [review]
Tested and everything worked as intended.

**Review by philgabr — DISMISSED** [review]
Works as intended.

**Review by EneaGore — DISMISSED** [review]
All seems to work as intented.

**Review by airelawaleria — DISMISSED** [review]
Re-approve

**Review by simonzilker — APPROVED** [review]
re-approve!

**Review by JohannesWeiss — APPROVED** [review]
Still works as intended on ts1. Re-Approve.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/VersionControlService.java [hunk 1]
```diff
@@ -131,10 +131,12 @@
      * @param sourceRepositoryName The name of the repository which should be copied
      * @param targetProjectKey The key of the target project to which to copy the new plan to
      * @param targetRepositoryName The desired name of the target repository
+     * @param targetPath The path where the repo is cloned into
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/VersionControlService.java:134]:
```suggestion
     * @param targetPath The path in the local Artemis server file system where the repo is cloned into in order to copy it
```

**Pull request merged**